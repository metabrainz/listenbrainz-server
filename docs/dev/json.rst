.. _json-doc:

JSON Documentation
==================

.. note:: Do not submit copyrighted information in these fields!

Submission JSON
---------------

To submit a listen via our API (see: :doc:`api`), ``POST`` a JSON document to
the ``submit-listens`` endpoint. Submit one of three types JSON documents:

- ``single``: Submit single listen

   - Indicates user just finished listening to track

   - ``payload`` should contain information about *exactly one* track

- ``playing_now``: Submit ``playing_now`` notification

   - Indicates that user just began listening to track

   - ``payload`` should contain information about *exactly one* track

   - Submitting ``playing_now`` documents is optional

   - Timestamp must be omitted from a ``playing_now`` submission.

- ``import``: Submit previously saved listens

   - ``payload`` should contain information about *at least one* track

   - submitting multiple listens in one go is allowed, but the complete JSON
     document may not exceed :data:`~webserver.views.api.MAX_LISTEN_SIZE` bytes
     in size

The ``listen_type`` element defines different types of submissions. The element
is placed at the top-most level of the JSON document. The only other required
element is the ``payload`` element. This provides an array of listens â€“ the
payload may be one or more listens (as designated by ``listen_type``)::

    {
      "listen_type": "single",
      "payload": [
          --- listen data here ---
      ]
    }

A sample listen payload may look like::

    {
      "listened_at": 1443521965,
      "track_metadata": {
        "additional_info": {
          "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
          "artist_mbids": [
            "db92a151-1ac2-438b-bc43-b82e149ddd50"
          ],
          "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
          "tags": [ "you", "just", "got", "rick rolled!"]
        },
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
        "release_name": "Whenever you need somebody"
      }
    }

A complete submit listen JSON document may look like::

    {
      "listen_type": "single",
      "payload": [
        {
          "listened_at": 1443521965,
          "track_metadata": {
            "additional_info": {
              "listening_from": "spotify",
              "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
              "artist_mbids": [
                "db92a151-1ac2-438b-bc43-b82e149ddd50"
              ],
              "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
              "tags": [ "you", "just", "got", "rick rolled!"]
            },
            "artist_name": "Rick Astley",
            "track_name": "Never Gonna Give You Up",
            "release_name": "Whenever you need somebody"
          }
        }
      ]
    }


Fetching listen JSON
--------------------

The JSON documents returned from our API look like the following::

    {
      "payload": {
        "count": 25,
        "user_id": "-- the MusicBrainz ID of the user --",
        "listens": [
          "-- listen data here ---"
        ]
      }
    }

The number of listens in the document are returned by the top-level ``count``
element. The ``user_id`` element contains the MusicBrainz ID of the user whose listens are
being returned. The other element is the ``listens`` element. This is a list which contains
the listen JSON elements (described above).

The JSON document returned by the API endpoint for getting tracks being played right now
is the same as above, except that it also contains the ``payload/playing_now`` element as a
boolean set to True.


Payload JSON details
--------------------

A minimal payload must include
``track_metadata/artist_name`` and ``track_metadata/track_name`` elements::

    {
      "track_metadata": {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      }
    }

``artist_name`` and ``track_name`` elements must be simple strings.

The payload will also include the ``listened_at`` element which must be an integer
representing the Unix time when the track was listened to. The only exception to this
rule is when the listen is being played right now and has been retrieved from the
endpoint to get listens being played right now. The ``listened_at`` element will be
absent for such listens.

Add additional metadata you may have for a track to the ``additional_info``
element. Any additional information allows us to better correlate your listen
data to existing MusicBrainz-based data. If you have MusicBrainz IDs available,
submit them!

The following optional elements may also be included in the ``track_metadata`` element:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``release_name``        the name of the release this recording was played from.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

The following optional elements may also be included in the ``additional_info`` element. If you do not have
the data for any of the following fields, omit the key entirely:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``artist_mbids``        A list of MusicBrainz Artist IDs, one or more Artist IDs may be included here. If you have a complete MusicBrainz artist credit that contains multiple Artist IDs, include them all in this list.
``release_group_mbid``  A MusicBrainz Release Group ID of the release group this recording was played from.
``release_mbid``        A MusicBrainz Release ID of the release this recording was played from.
``recording_mbid``      A MusicBrainz Recording ID of the recording that was played.
``track_mbid``          A MusicBrainz Track ID associated with the recording that was played.
``work_mbids``          A list of MusicBrainz Work IDs that may be associated with this recording.
``tracknumber``         The tracknumber of the recording. This first recording on a release is tracknumber 1.
``isrc``                The ISRC code associated with the recording.
``spotify_id``          The Spotify track URL associated with this recording.  e.g.: http://open.spotify.com/track/1rrgWMXGCGHru5bIRxGFV0
``tags``                A list of user defined tags to be associated with this recording. These tags are similar to last.fm tags. For example, you have apply tags such as ``punk``, ``see-live``, ``smelly``. You may submit up to :data:`~webserver.views.api.MAX_TAGS_PER_LISTEN` tags and each tag may be up to :data:`~webserver.views.api.MAX_TAG_SIZE` characters large.
``listening_from``       The source of the listen, i.e the name of the client or service which submits the listen.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

At this point, we are not scrubbing any superflous elements that may be
submitted via the ``additional_info`` element. We're open to see how people
will make use of these unspecified fields and may decide to formally specify or
scrub elements in the future.
