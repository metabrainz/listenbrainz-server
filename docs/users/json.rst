.. _json-doc:

JSON Documentation
==================

.. note:: Do not submit copyrighted information in these fields!

Submission JSON
---------------

To submit a listen via our API (see: :doc:`api/core`), ``POST`` a JSON document to
the ``submit-listens`` endpoint. Submit one of three types JSON documents:

- ``single``: Submit single listen

   - Indicates user just finished listening to track

   - ``payload`` should contain information about *exactly one* track

- ``playing_now``: Submit ``playing_now`` notification

   - Indicates that user just began listening to track

   - ``payload`` should contain information about *exactly one* track

   - Submitting ``playing_now`` documents is optional

   - Timestamp must be omitted from a ``playing_now`` submission.

.. note::

    Playing Now listens are only stored temporarily. A playing now listen must be
    submitted again as a ``single`` or ``import`` for permanent storage.


- ``import``: Submit previously saved listens

   - ``payload`` should contain information about *at least one* track

   - submitting multiple listens in one go is allowed, but the complete JSON
     document may not exceed :data:`~webserver.views.api.MAX_LISTEN_SIZE` bytes
     in size

The ``listen_type`` element defines different types of submissions. The element
is placed at the top-most level of the JSON document. The only other required
element is the ``payload`` element. This provides an array of listens â€“ the
payload may be one or more listens (as designated by ``listen_type``):

.. code-block:: json

    {
      "listen_type": "single",
      "payload": [
          "--- listen data here ---"
      ]
    }

A sample listen payload may look like:

.. code-block:: json

    {
      "listened_at": 1443521965,
      "track_metadata": {
        "additional_info": {
          "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
          "artist_mbids": [
            "db92a151-1ac2-438b-bc43-b82e149ddd50"
          ],
          "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
          "tags": [ "you", "just", "got", "rick rolled!"]
        },
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
        "release_name": "Whenever you need somebody"
      }
    }

A complete submit listen JSON document may look like:

.. code-block:: json

    {
      "listen_type": "single",
      "payload": [
        {
          "listened_at": 1443521965,
          "track_metadata": {
            "additional_info": {
              "media_player": "Rhythmbox",
              "submission_client": "Rhythmbox ListenBrainz Plugin",
              "submission_client_version": "1.0",
              "release_mbid": "bf9e91ea-8029-4a04-a26a-224e00a83266",
              "artist_mbids": [
                "db92a151-1ac2-438b-bc43-b82e149ddd50"
              ],
              "recording_mbid": "98255a8c-017a-4bc7-8dd6-1fa36124572b",
              "tags": [ "you", "just", "got", "rick rolled!"]
            },
            "artist_name": "Rick Astley",
            "track_name": "Never Gonna Give You Up",
            "release_name": "Whenever you need somebody"
          }
        }
      ]
    }


Fetching listen JSON
--------------------

The JSON documents returned from our API look like the following:

.. code-block:: json

    {
      "payload": {
        "count": 25,
        "user_id": "-- the MusicBrainz ID of the user --",
        "listens": [
          "-- listen data here ---"
        ]
      }
    }

The number of listens in the document are returned by the top-level ``count``
element. The ``user_id`` element contains the MusicBrainz ID of the user whose listens are
being returned. The other element is the ``listens`` element. This is a list which contains
the listen JSON elements (described above).

The JSON document returned by the API endpoint for getting tracks being played right now
is the same as above, except that it also contains the ``payload/playing_now`` element as a
boolean set to True.


Payload JSON details
--------------------

A minimal payload must include
``track_metadata/artist_name`` and ``track_metadata/track_name`` elements:

.. code-block:: json

    {
      "track_metadata": {
        "artist_name": "Rick Astley",
        "track_name": "Never Gonna Give You Up",
      }
    }

``artist_name`` and ``track_name`` elements must be simple strings.

The payload should also include the ``listened_at`` element, which must be an integer
representing the Unix time when the track was listened to. The minimum accepted
value for this field is :data:`~webserver.views.api.LISTEN_MINIMUM_TS`.
playing_now requests should not have a ``listened_at`` field

Add additional metadata you may have for a track to the ``additional_info``
element. Any additional information allows us to better correlate your listen
data to existing MusicBrainz-based data. If you have MusicBrainz IDs available,
submit them!

The following optional elements may also be included in the ``track_metadata`` element:

======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
element                 description
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================
``release_name``        the name of the release this recording was played from.
======================= ===========================================================================================================================================================================================================================================================================================================================================================================================================

The following optional elements may also be included in the ``additional_info`` element. If you do not have
the data for any of the following fields, omit the key entirely:

.. list-table:: Title
   :widths: 25 50
   :header-rows: 1

   * - element
     - description
   * - ``artist_mbids``
     - A list of MusicBrainz Artist IDs, one or more Artist IDs may be included here. If you have a complete MusicBrainz artist credit that contains multiple Artist IDs, include them all in this list.
   * - ``release_group_mbid``
     - A MusicBrainz Release Group ID of the release group this recording was played from.
   * - ``release_mbid``
     - A MusicBrainz Release ID of the release this recording was played from.
   * - ``recording_mbid``
     - A MusicBrainz Recording ID of the recording that was played.
   * - ``track_mbid``
     - A MusicBrainz Track ID associated with the recording that was played.
   * - ``work_mbids``
     - A list of MusicBrainz Work IDs that may be associated with this recording.
   * - ``tracknumber``
     - The tracknumber of the recording. This first recording on a release is tracknumber 1.
   * - ``isrc``
     - The ISRC code associated with the recording.
   * - ``spotify_id``
     - The Spotify track URL associated with this recording.  e.g.: http://open.spotify.com/track/1rrgWMXGCGHru5bIRxGFV0
   * - ``tags``
     - A list of user-defined folksonomy tags to be associated with this recording. For example, you have apply tags such as ``punk``, ``see-live``, ``smelly``. You may submit up to :data:`~webserver.views.api.MAX_TAGS_PER_LISTEN` tags and each tag may be up to :data:`~webserver.views.api.MAX_TAG_SIZE` characters large.
   * - ``media_player``
     - The name of the program being used to listen to music. Don't include a version number here.
   * - ``media_player_version``
     - The version of the program being used to listen to music.
   * - ``submission_client``
     - The name of the client that is being used to submit listens to ListenBrainz. If the media player has the ability to submit listens built-in then this value may be the same as ``media_player``. Don't include a version number here.
   * - ``submission_client_version``
     - The version of the submission client.
   * - ``music_service``
     - If the song being listened to comes from an online service, the canonical domain of this service (see below for more details).
   * - ``music_service_name``
     - If the song being listened to comes from an online service and you don't know the canonical domain, a name that represents the service.
   * - ``origin_url``
     - If the song of this listen comes from an online source, the URL to the place where it is available. This could be a spotify url (see ``spotify_id``), a YouTube video URL, a Soundcloud recording page URL, or the full URL to a public MP3 file. If there is a webpage for this song (e.g. Youtube page, Soundcloud page) **do not** try and resolve the URL to an actual audio resource.

.. note::

  **Music service names**

  The ``music_service`` field should be a domain name rather than a textual description or URL. This allows us to refer unambiguously to a service without worrying
  about capitalization or full/short names (such as the difference between "Internet Archive", "The Internet Archive" or "Archive").
  If we use this data on ListenBrainz, we will perform a mapping from the domain name to a canonical name. Below is an example of mappings that we currently support.
  If you are submitting from a service which doesn't appear in this list, you should determine a canonical domain from the domain of the service.
  Only if you cannot determine a domain for the service should you use the text-only ``music_service_name`` field.

  .. list-table:: Music services domain/name mapping
     :widths: 25 50
     :header-rows: 1

     * - domain
       - name
     * - ``spotify.com``
       - Spotify
     * - ``bandcamp.com``
       - Bandcamp
     * - ``youtube.com``
       - YouTube
     * - ``music.youtube.com``
       - YouTube Music
     * - ``deezer.com``
       - Deezer
     * - ``tidal.com``
       - TIDAL
     * - ``music.apple.com``
       - Apple Music
     * - ``archive.org``
       - Internet Archive
     * - ``soundcloud.com``
       - Soudcloud
     * - ``jamendo.com``
       - Jamendo Music
     * - ``play.google.com``
       - Google Play Music


Client Metadata examples
------------------------

Here are a few examples of how to fill in the ``media_player``, ``submission_client`` and ``music_service`` fields based on our
current recommendations.

BrainzPlayer on the ListenBrainz website playing a video from YouTube
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "BrainzPlayer",
            "music_service": "youtube.com",
            "origin_url": "https://www.youtube.com/watch?v=JKFBiaoFHoY",
            "submission_client": "BrainzPlayer"
        },
        "artist_name": "Mdou Moctar",
        "release_name": "Ilana (The Creator)",
        "track_name": "Inizgam"
    }
  }

BrainzPlayer on the ListenBrainz website playing a video from Spotify
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Note that even though the ``origin_url`` is ``https://open.spotify.com``, we set ``music_service``
to spotify.com (see above note).

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "BrainzPlayer",
            "music_service": "spotify.com",
            "origin_url": "https://open.spotify.com/track/5fEjp2F0Sqr9fMuLSaDqz0",
            "submission_client": "BrainzPlayer"
        },
        "artist_name": "Les Filles de Illighadad",
        "release_name": "Eghass Malan",
        "track_name": "Inssegh Inssegh"
    }
  }


Using Otter for Funkwhale on android, and submitting with Simple Scrobbler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In this case, the media player and submission client are completely separate programs. Because music is being played
from a user's private collection and not a streaming service, don't include music_service or origin_url.

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "Otter",
            "media_player_version": "1.0.21",
            "submission_client": "Simple Scrobbler"
            "submission_client_version": "1.7.0"
        },
        "artist_name": "Les Filles de Illighadad",
        "release_name": "Eghass Malan",
        "track_name": "Inssegh Inssegh"
    }
  }


Rhythmbox player listening to Jamendo
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: JSON

  {
    "track_metadata": {
        "additional_info": {
            "media_player": "Rhythmbox",
            "music_service": "jamendo.com",
            "music_service_name": "Jamendo Music"
            "origin_url": "https://www.jamendo.com/track/1466090/universal-funk",
            "submission_client": "Rhythmbox ListenBrainz Plugin"
        },
        "artist_name": "Duo Teslar",
        "track_name": "Universal Funk"
    }
  }

Listening to a recording from Bandcamp and submitting with the browser extension WebScrobbler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Because playback happens in the browser, there is no specific ``media_player``.

.. code-block:: JSON

  {
	"track_metadata": {
		"additional_info": {
			"music_service": "bandcamp.com",
			"music_service_name": "Bandcamp",
			"submission_client": "WebScrobbler",
			"submission_client_version": "v2.48.0"
			"origin_url": "https://greencookierecords.bandcamp.com/track/shake",
		},
		"artist_name": "I Mitomani Beat",
		"release_name": "Fuori Dal Tempo",
		"track_name": "Shake",
	}
  }

At this point, we are not removing any other elements that may be
submitted via the ``additional_info`` element. We're open to see how people
will make use of these unspecified fields and may decide to formally specify or
scrub elements in the future.
