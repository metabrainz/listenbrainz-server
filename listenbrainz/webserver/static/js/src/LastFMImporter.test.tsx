import * as React from "react";
import { mount, shallow } from "enzyme";
import LastFmImporter from "./LastFMImporter";
// Mock data to test functions
import * as page from "./__mocks__/page.json";
import * as getInfo from "./__mocks__/getInfo.json";
import * as getInfoNoPlayCount from "./__mocks__/getInfoNoPlayCount.json";
// Output for the mock data
import * as encodeScrobbleOutput from "./__mocks__/encodeScrobbleOutput.json";

jest.useFakeTimers();
const props = {
  user: {
    id: "id",
    name: "dummyUser",
    auth_token: "foobar",
  },
  profileUrl: "http://profile",
  apiUrl: "apiUrl",
  lastfmApiUrl: "http://ws.audioscrobbler.com/2.0/",
  lastfmApiKey: "foobar",
};

describe("encodeScrobbles", () => {
  it("encodes the given scrobbles correctly", () => {
    expect(LastFmImporter.encodeScrobbles(page)).toEqual(encodeScrobbleOutput);
  });
});

let instance: LastFmImporter;

describe("getNumberOfPages", () => {
  beforeEach(() => {
    const wrapper = shallow<LastFmImporter>(<LastFmImporter {...props} />);
    instance = wrapper.instance();
    instance.setState({ lastfmUsername: "dummyUser" });
    // Mock function for fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(page),
      });
    });
  });

  it("should call with the correct url", () => {
    instance.getNumberOfPages();

    expect(window.fetch).toHaveBeenCalledWith(
      `${props.lastfmApiUrl}?method=user.getrecenttracks&user=${instance.state.lastfmUsername}&api_key=${props.lastfmApiKey}&from=1&format=json`
    );
  });

  it("should return number of pages", async () => {
    const num = await instance.getNumberOfPages();
    expect(num).toBe(1);
  });

  it("should return -1 if there is an error", async () => {
    // Mock function for failed fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: false,
      });
    });

    const num = await instance.getNumberOfPages();
    expect(num).toBe(-1);
  });
});

describe("getTotalNumberOfScrobbles", () => {
  beforeEach(() => {
    const wrapper = shallow<LastFmImporter>(<LastFmImporter {...props} />);
    instance = wrapper.instance();
    instance.setState({ lastfmUsername: "dummyUser" });
    // Mock function for fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(getInfo),
      });
    });
  });

  it("should call with the correct url", () => {
    instance.getTotalNumberOfScrobbles();

    expect(window.fetch).toHaveBeenCalledWith(
      `${props.lastfmApiUrl}?method=user.getinfo&user=${instance.state.lastfmUsername}&api_key=${props.lastfmApiKey}&format=json`
    );
  });

  it("should return number of pages", async () => {
    const num = await instance.getTotalNumberOfScrobbles();
    expect(num).toBe(1026);
  });

  it("should return -1 if playcount is not available", async () => {
    // Mock function for fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(getInfoNoPlayCount),
      });
    });

    const num = await instance.getTotalNumberOfScrobbles();
    expect(num).toBe(-1);
  });

  it("should throw an error when fetch fails", async () => {
    // Mock function for failed fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: false,
      });
    });
    await expect(instance.getTotalNumberOfScrobbles()).rejects.toThrowError();
  });
  it("should show the error message in importer", async () => {
    // Mock function for failed fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: false,
      });
    });
    await expect(instance.getTotalNumberOfScrobbles()).rejects.toThrowError();
    expect(instance.state.msg?.props.children).toMatch(
      "An error occurred, please try again. :("
    );
  });
});

describe("getPage", () => {
  beforeEach(() => {
    const wrapper = shallow<LastFmImporter>(<LastFmImporter {...props} />);
    instance = wrapper.instance();
    instance.setState({ lastfmUsername: "dummyUser" });
    // Mock function for fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve(page),
      });
    });
  });

  it("should call with the correct url", () => {
    instance.getPage(1);

    expect(window.fetch).toHaveBeenCalledWith(
      `${props.lastfmApiUrl}?method=user.getrecenttracks&user=${instance.state.lastfmUsername}&api_key=${props.lastfmApiKey}&from=1&page=1&format=json`
    );
  });

  it("should call encodeScrobbles", async () => {
    // Mock function for encodeScrobbles
    LastFmImporter.encodeScrobbles = jest.fn(() => ["foo", "bar"]);

    const data = await instance.getPage(1);
    expect(LastFmImporter.encodeScrobbles).toHaveBeenCalledTimes(1);
    expect(data).toEqual(["foo", "bar"]);
  });

  it("should retry if 50x error is recieved", async () => {
    // Mock function for fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: false,
        status: 503,
      });
    });

    await instance.getPage(1);
    // There is no direct way to check if retry has been called
    expect(setTimeout).toHaveBeenCalledTimes(1);

    jest.runAllTimers();
  });

  it("should skip the page if 40x is recieved", async () => {
    // Mock function for failed fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: false,
        status: 404,
      });
    });

    await instance.getPage(1);
    expect(setTimeout).not.toHaveBeenCalled();
  });

  it("should retry if there is any other error", async () => {
    // Mock function for fetch
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => Promise.reject(),
      });
    });

    await instance.getPage(1);
    // There is no direct way to check if retry has been called
    expect(setTimeout).toHaveBeenCalledTimes(1);
    jest.runAllTimers();
  });
});

describe("submitPage", () => {
  beforeEach(() => {
    const wrapper = shallow<LastFmImporter>(<LastFmImporter {...props} />);
    instance = wrapper.instance();
    instance.setState({ lastfmUsername: "dummyUser" });
    instance.getRateLimitDelay = jest.fn().mockImplementation(() => 0);
    instance.updateRateLimitParameters = jest.fn();
  });

  it("calls submitListens once", async () => {
    // window.fetch = jest.fn().mockImplementation(() => {
    //   return Promise.resolve({
    //     ok: true,
    //     json: () => Promise.resolve({ status: 200 }),
    //   });
    // });
    instance.APIService.submitListens = jest.fn().mockImplementation(() => {
      return Promise.resolve({ status: 200 });
    });
    instance.submitPage([
      {
        listened_at: 1000,
        track_metadata: {
          artist_name: "foobar",
          track_name: "bazfoo",
        },
      },
    ]);

    jest.runAllTimers();

    // Flush all promises
    // https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises
    await new Promise((resolve) => setImmediate(resolve));

    expect(instance.APIService.submitListens).toHaveBeenCalledTimes(1);
  });

  it("calls updateRateLimitParameters once", async () => {
    instance.APIService.submitListens = jest.fn().mockImplementation(() => {
      return Promise.resolve({ status: 200 });
    });
    instance.submitPage([
      {
        listened_at: 1000,
        track_metadata: {
          artist_name: "foobar",
          track_name: "bazfoo",
        },
      },
    ]);

    jest.runAllTimers();

    // Flush all promises
    // https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises
    await new Promise((resolve) => setImmediate(resolve));

    expect(instance.updateRateLimitParameters).toHaveBeenCalledTimes(1);
    expect(instance.updateRateLimitParameters).toHaveBeenCalledWith({
      status: 200,
    });
  });

  it("calls getRateLimitDelay once", async () => {
    instance.submitPage([
      {
        listened_at: 1000,
        track_metadata: {
          artist_name: "foobar",
          track_name: "bazfoo",
        },
      },
    ]);
    expect(instance.getRateLimitDelay).toHaveBeenCalledTimes(1);
  });
});

describe("LastFmImporter Page", () => {
  it("renders", () => {
    const wrapper = mount(<LastFmImporter {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("modal renders when button clicked", () => {
    const wrapper = shallow(<LastFmImporter {...props} />);
    // Simulate submiting the form
    wrapper.find("form").simulate("submit", {
      preventDefault: () => null,
    });

    // Test if the show property has been set to true
    expect(wrapper.exists("LastFMImporterModal")).toBe(true);
  });

  it("submit button is disabled when input is empty", () => {
    const wrapper = shallow(<LastFmImporter {...props} />);
    // Make sure that the input is empty
    wrapper.setState({ lastfmUsername: "" });

    // Test if button is disabled
    expect(wrapper.find('input[type="submit"]').props().disabled).toBe(true);
  });
});

describe("LastFmImporter Page", () => {
  it("should properly convert latest imported timestamp to string", () => {
    // Check getlastImportedString() and formatting
    let testDate = Number(page["recenttracks"]["track"][0]["date"]["uts"]);
    let lastImportedDate = new Date(testDate * 1000);
    let msg = lastImportedDate.toLocaleString("en-US", {
        month: "short",
        day: "2-digit",
        year: "numeric",
        hour: "numeric",
        minute: "numeric",
        hour12: true,
    });

    expect(instance.getlastImportedString(testDate)).toMatch(msg);
    expect(instance.getlastImportedString(testDate).length).not.toEqual(0);
  });
});