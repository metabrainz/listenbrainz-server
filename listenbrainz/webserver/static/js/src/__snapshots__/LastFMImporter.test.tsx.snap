// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastFmImporter Page renders 1`] = `"<div class=\\"Importer\\"><form><input type=\\"text\\" name=\\"lastfmUsername\\" placeholder=\\"Last.fm Username\\" size=\\"30\\" value=\\"\\"><input type=\\"submit\\" disabled=\\"\\" value=\\"Import Now!\\"></form></div>"`;
